gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/testing/KspIntegrationTestRule.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2020 Google LLC
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.devtools.ksp.gradle.testing
+
+import com.google.devtools.ksp.processing.SymbolProcessor
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.rules.TemporaryFolder
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+import kotlin.reflect.KClass
+
+/**
+ * JUnit test rule to setup a [TestProject] which contains a KSP processor module and an
+ * application. The application can either be an android app or jvm app.
+ * Test must call [setupAppAsAndroidApp] or [setupAppAsJvmApp] before using the [runner].
+ */
+class KspIntegrationTestRule(
+    private val tmpFolder: TemporaryFolder
+) : TestWatcher() {
+    /**
+     * Initialized when the test starts.
+     */
+    private lateinit var testProject: TestProject
+
+    /**
+     * The application module in the test project
+     */
+    val appModule
+        get() = testProject.appModule
+
+    /**
+     * The processor module in the test project
+     */
+    val processorModule
+        get() = testProject.processorModule
+
+    /**
+     * The configuration passed from the KSP's main build which includes important setup information
+     * like KSP version, local maven repo etc.
+     */
+    val testConfig = TestConfig.read()
+
+    /**
+     * Returns a gradle runner that is ready to run tasks on the test project.
+     */
+    fun runner(): GradleRunner {
+        testProject.writeFiles()
+        return GradleRunner.create()
```

Comment:
To make sure tests can use `maxParallelForks` property i.e run in parallel, I'd specify GRADLE_USER home env var. Each for gets an id from `System.getProperty("org.gradle.test.worker")` and each for can get the gradle user home location as "$GRADLE_HOME_FOR_TEST/$workerId".

Reply:
i'm running these tests in process (for debuggability) and it is not compatible w/ environment variable passing.

Reply:
I guess that means you are unable to run them in parallel, but that may be ok for now.